### DISCUSSION TITLE
Feature request: Vector embeddings similarity search

### DISCUSSION THREAD

**Discussion by jp200k** [duckdb/duckdb#10520] [Ideas]
I discovered that Microsoft Semantic Kernel includes DuckDB as one of the “Available connectors to vector databases” 1. Intrigued, I dedicated several days to studying DuckDB with high hopes. However, I was disappointed to find that there are currently no functions available for vector embeddings similarity search in DuckDB. If such a feature were added, it would significantly enhance DuckDB’s value.

**Comment by carlopi** [comment]
You can look at the ARRAY type: https://duckdb.org/docs/sql/data_types/array and it's functions.
It's available on nightlies built AND will be part of the upcoming v0.10.0 release

**Comment by Giorgi** [comment]
There are functions already for similarity in DuckDB: `list_cosine_similarity` `list_distance`

and the Semantic Kernel DuckDB connector is using it so you shouldn't be disappointed.

**Comment by FrancoisLepoutre** [comment]
Thanks you all for starting and feeding this thread!  I have not dived into semantic search but intend to.

I understand that Duckdb is no "semantic" engine and that there is no intention to provide such functionality. With good reason. But I just grasped from what Giorgi and Carlo mentioned that:
1. our beloved waterfowl already offers some support for list-distance-similarity calculation,
2. the incoming "array" data type could reduce the cpu handling costs for these "distance operations".

If you work in the python data ecosystem, am I wrong to assume that most building blocks required to produce basic duckdb-based or duckdb-fed semantic search may already be at hand? Or am I missing some important stuff?
